File: porn-blocker/app/src/main/java/com/pornblocker/SimpleBrowser.java
package com.pornblocker

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment

public class SimpleBrowser extends Fragment {

    public static SimpleBrowser newInstance() {
        return new SimpleBrowser()
    }

    private WebView webView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        WebView wv = new WebView(requireContext());
        this.webView = wv;
        return wv;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setDomStorageEnabled(true);
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(android.webkit.WebView view, String url) {
                if (url != null && new BlocklistManager(view.getContext()).isBlocked(url)) {
                    view.loadData(
                        "<h1>Blocked</h1><p>This site is blocked by your settings.</p>",
                        "text/html",
                        "UTF-8"
                    );
                }
            }
        });
        webView.loadData("<h1>Ready</h1>", "text/html", "UTF-8");
    }

    public void load(String url) {
        if (webView != null) webView.loadUrl(url);
    }

    @Override
    public void onDestroyView() {
        if (webView != null) {
            webView.setWebViewClient(new WebViewClient());
            webView.stopLoading();
            webView.loadUrl("about:blank");
            webView.clearHistory();
            webView.removeAllViews();
            webView.destroy();
            webView = null;
        }
        super.onDestroyView();
    }
}

File: porn-blocker/app/src/main/java/com/pornblocker/SimpleBrowser.kt
package com.pornblocker

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment

class SimpleBrowser : Fragment() {

    companion object {
        fun newInstance() = SimpleBrowser()
    }

    private var webView: WebView? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return WebView(requireContext()).also { webView = it }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        webView?.apply {
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true
            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView?, url: String?) {
                    if (url != null && (BlocklistManager(view!!.context)).isBlocked(url)) {
                        loadData(
                            "<h1>Blocked</h1><p>This site is blocked by your settings.</p>",
                            "text/html",
                            "UTF-8"
                        )
                    }
                }
            }
            loadData("<h1>Ready</h1>", "text/html", "UTF-8")
        }
    }

    fun load(url: String) {
        webView?.loadUrl(url)
    }

    override fun onDestroyView() {
        webView?.apply {
            webViewClient = WebViewClient()
            stopLoading()
            loadUrl("about:blank")
            clearHistory()
            removeAllViews()
            destroy()
        }
        webView = null
        super.onDestroyView()
    }
}
